lib/components/tag/tag.dart

```diff
@@ -0,0 +1,213 @@
+  void Function() onDelete = () {};
+  void Function() onSelect = () {};
+  void Function() onAccept = () {};
+
+  TagView(this._text, this._tagId,
+      {String groupId,
+      bool selectable = true,
+      bool editable = false,
+      bool editableOnAdd = false,
+      bool removable = false,
+      bool acceptable = false,
+      bool suggested = false,
+      TagStyle tagStyle = TagStyle.None,
+      bool disableDoubleClickEdit = false}) {
+    _selectable = selectable;
+    _editable = editable;
+    _editableOnAdd = editableOnAdd;
+    _removable = removable;
+    _acceptable = acceptable;
+    _suggested = suggested;
+
+    renderElement = SpanElement()
+      ..dataset['id'] = _tagId
+      ..dataset['group-id'] = groupId ?? ''
+      ..classes.add('tag');
+
+    _tagText = SpanElement()
+      ..classes.add('tag-text')
+      ..dataset['placeholder'] = "untitled tag";
+    _tagActions = SpanElement()..classes.add('tag-actions');
+
+    _tagText.innerText = this._text;
+    if (suggested) {
+      var suggestedIcon = SpanElement()..className = 'fas fa-robot tag-suggested__icon';
+      _tagText.append(suggestedIcon);
+      _tagText.classes.toggle('tag-text--suggested', true);
+    }
+
+    updateStyle(tagStyle);
+
+    if (_selectable) {
+      _tagText.onClick.listen((_) {
+        onSelect();
+      });
+    }
+
+    if (_acceptable) {
+      _acceptButton = Button(ButtonType.confirm, onClick: (_) {
+        _acceptTag();
+      }).renderElement;
+      _tagActions.append(_acceptButton);
+    }
+
+    if (_editable) {
+      _editButton = Button(ButtonType.edit, onClick: (_) {
+        _makeEditable();
+      }).renderElement;
+      _tagActions.append(_editButton);
+
+      _confirmButton = Button(ButtonType.confirm, onClick: (_) {
+        _confirmEdit();
+      }).renderElement;
+
+      _cancelButton = Button(ButtonType.cancel, onClick: (_) {
+        _cancelEditing();
+      }).renderElement;
+
+      if (!disableDoubleClickEdit) {
+        _tagText.onDoubleClick.listen((_) {
+          _makeEditable();
+        });
+      }
+    }
+
+    if (_removable) {
+      _deleteButton = Button(ButtonType.remove, onClick: (_) {
+        _deleteTag();
+      }).renderElement;
+      _tagActions.append(_deleteButton);
+    }
+
+    if (editable && editableOnAdd) {
+      _makeEditable();
+    }
+
+    renderElement..append(_tagText)..append(_tagActions);
+  }
+
+  void _makeEditable() {
+    _tagText
+      ..contentEditable = "true"
+      ..focus()
+      ..onKeyDown.listen((event) {
+        if (event.keyCode == KeyCode.ENTER) {
+          event.preventDefault();
+          _confirmEdit();
+        }
+        if (event.keyCode == KeyCode.ESC) {
+          event.preventDefault();
+          _cancelEditing();
+        }
+      });
+    renderElement.classes.toggle("tag-text--editing", true);
+
+    _tagActions.children.clear();
+    _tagActions..append(_confirmButton)..append(_cancelButton);
+  }
+
+  void _cancelEditing() {
```

Comment:
can we make this a public method? I think there might be some use in exposing this as part of a Tag component's API to be able to support more complex interaction (e.g. in Nook, if you're editing a tag, it will be cancelled if you switch conversations). I think `beginEdit` (rename of `_makeEditable`), `cancelEdit` (rename of `_cancelEditing`), `confirmEdit`, `deleteTag` and `acceptTag` should all be public.

Reply:
one further thought - if we have `beginEdit`, then we could add the behaviour of `editableOnAdd` by: creating the tag, with `editable = false`, and then calling immediately after `beginEdit` - is that right? If so, perhaps we can remove the `editableOnAdd` attribute, which would simplify a bit the number of parameters available.

Maybe I'm getting excited without good reason, but I think it's quite helpful for the UI components to have their own behaviour, but also they can be asked to do something by the logic of the UI, and this feels to me like a good example where that kind of pattern works well. This allows the UI to know enough, but not all, and keeping the logic of handling more special cases outside of the UI library.